In [2]:
# Competition-Solution/notebooks/trial/00_data_profiling/00_data_profiling_basic.ipynb

### Data Profiling - Trial Dataset

The notebooks within the **'00_data_profiling'** folder perform initial exploratory data analysis on the trial dataset, focusing on the three subtasks:

- **Subtask 1:** Call2Action

- **Subtask 2:** Attacks on the Democratic Basic Order (FDGO)

- **Subtask 3:** Violence Detection


Data Profiling is performed in order to get a better understanding of the data we are working with. When data profiling, we are looking to gather a high-level overview and statistical summary of our dataset.
In essence, it is a report-card for our data.

The insights from this data 'report-card" will help guide the approach taken for each subtask. 
Additionaly, it will provide a insights for retrospectively analyzing our models" classification patterns and the factors influencing them.

---

##### <b>Imports</b>

In [3]:
import sys
import pandas as pd
from rich.console import Console

sys.path.append("../../../src")
import data_utils

console = Console()

##### <b>Loading the data</b>

In [4]:
df_call2action, df_dbo, df_violence = data_utils.load_competition_data(data_dir="../../../data/raw/", type="trial")

Loading competition data...

Data of type 'trial' loaded successfully.

##### <b>Initial Head and Tail</b>

* ##### C2A Head & Tail

In [5]:
df_call2action.head()

,id,description,C2A
0,982640125107744,Das sind unsere Facharbeiter und Ärzte !!! Na ...,False
1,1012237782147978,Man kann den Dünnschiss nicht mehr ersehen!,False
2,1010223125682777,"Hut ab, Frau Festerling! Ich unterschreibe jed...",False
3,876541075717650,Tod nach german,False
4,1035119656526457,Ich verneige mich,False


In [6]:
df_call2action.tail()

,id,description,C2A
1046,1120166181355137,sehr gut geschrieben,False
1047,946052645433159,"Ja was denn, glaubt Du, die reisen dafür nach ...",False
1048,1065801736791582,Absolut richtig was da passiert. .von Politike...,False
1049,861943823844042,"Pegida wird sich genauso auflösen wie Leipzig,...",False
1050,1134146769957078,Reicht wenn du dumme schnöpfe die mit heim nimmst,False


* ##### DBO Head & Tail

In [7]:
df_dbo.head()

,id,description,DBO
0,861943053844119,Besser als auf sone drecks Veranstaltung zu ge...,nothing
1,1012568078781615,"Roland Kauß Ist mir nicht peinlich, wüßte nich...",nothing
2,1152526151452473,Ich auch....im Moment rennen sie ja alle wie v...,nothing
3,1134073499964405,Klar bimbo nach dem Wahlsieg trumps,nothing
4,862087190496372,"Karla,wir lieben dich??...",nothing


In [8]:
df_dbo.tail()

,id,description,DBO
1048,982696511768772,Frau Merkel sie kommen jetzt Alle!,nothing
1049,882737588431332,Nur einer ... Sinnlos !!! ??,nothing
1050,1095908543780901,Was wollen die? Die sind doch in Sicherheit!,nothing
1051,982670668438023,"Gio, das ist nicht in einem Satz zu erklären. ...",criticism
1052,1007240065981083,"der Steinmeier ist doch bloß ein Einschleimer,...",criticism


* ##### VIO Head & Tail

In [9]:
df_violence.head()

,id,description,VIO
0,1096028327102256,wenn es denen nicht passt: raus mit den flücht...,False
1,876327089072382,"Deutsche werden weniger, dafür Deutschland Bun...",False
2,882678818437209,Unser Auslandskorrespondent Hung Nong bestätig...,False
3,1043827455655677,Hat jemand gerade den Schauspieler Axel Milber...,False
4,876291782409246,Diesen Eindruck vermitteln nur unsere Politike...,False


In [10]:
df_violence.tail()

,id,description,VIO
1046,991329500905473,Hut ab vor jedem einzelnen von EUCH ihr seid d...,False
1047,1007131215991968,De supermarkt in Duitsland zijn problemen onts...,False
1048,982640951774328,Die Einschläge kommen immer näher??,False
1049,1038628736175549,"Falls Straches Bericht stimmt, so ist dies ein...",False
1050,1095986027106486,Warum unternimmt die Polizei hier nichts?,False


##### <b>Row (Samples) and Column (Features) Counts</b>

In [11]:
# Dataset Shapes
c2a_shape = (df_call2action.shape[0], df_call2action.shape[1])
dbo_shape = (df_dbo.shape[0], df_dbo.shape[1])
violence_shape = (df_violence.shape[0], df_violence.shape[1])

total_rows = c2a_shape[0] + dbo_shape[0] + violence_shape[0]


print(f"Subtask 1 (Call2Action): {c2a_shape[0]} rows, {c2a_shape[1]} columns")
print(f"Subtask 2 (DBO): {dbo_shape[0]} rows, {dbo_shape[1]} columns")
print(f"Subtask 3 (Violence): {violence_shape[0]} rows, {violence_shape[1]} columns")

print(f"Total rows: {total_rows}")

Subtask 1 (Call2Action): 1051 rows, 3 columns
Subtask 2 (DBO): 1053 rows, 3 columns
Subtask 3 (Violence): 1051 rows, 3 columns
Total rows: 3155


<u><p id="row-column-counts">Interpretation</p></u>
- <b>Call2Action</b> has 1051 rows and three columns<br>
  * Call2Action has 2 classes (TRUE, FALSE)
- <b>DBO</b> has 1053 rows and three columns<br>
  * DBO has 4 classes (nothing, criticism, agitation, subversive)
- <b>Violence</b> has 1051 rows and three columns<br>
  * Violence has 2 classes (TRUE, FALSE)
<br>
<br>
- Total rows: 3155

##### <b>Data Types</b>

In [13]:
print(f"Subtask 1 (Call2Action): {df_call2action.dtypes}")
print(f"Subtask 2 (DBO): {df_dbo.dtypes}")
print(f"Subtask 3 (Violence): {df_violence.dtypes}")

Subtask 1 (Call2Action): id              int64
description    object
C2A              bool
dtype: object
Subtask 2 (DBO): id              int64
description    object
DBO            object
dtype: object
Subtask 3 (Violence): id              int64
description    object
VIO              bool
dtype: object


<u><p>Interpretation</p></u>
- Subtask 1 (Call2Action):
  - `id`: int64
  - `description`: object
  - `C2A`: bool
- Subtask 2 (DBO):
  - `id`: int64
  - `description`: object
  - `DBO`: object
- Subtask 3 (Violence):
  - `id`: int64
  - `description`: object
  - `VIO`: bool

Note: 'object' in this case is simply a string.

##### <b>Missing/Null Values</b>

In [ ]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action.isnull().sum())

print(f"Subtask 2 (DBO):")
print(df_dbo.isnull().sum())

print(f"Subtask 3 (Violence):")
print(df_violence.isnull().sum())

Subtask 1 (Call2Action):
id             0
description    0
C2A            0
dtype: int64
Subtask 2 (DBO):
id             0
description    0
DBO            0
dtype: int64
Subtask 3 (Violence):
id             0
description    0
VIO            0
dtype: int64


<u><p>Interpretation</p></u>
* No missing/null values!

##### <b>Duplicate Values within the individual subtask datasets</b>

In [17]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action.duplicated().sum())

print(f"Subtask 2 (DBO):")
print(df_dbo.duplicated().sum())

print(f"Subtask 3 (Violence):")
print(df_violence.duplicated().sum())

Subtask 1 (Call2Action):
0
Subtask 2 (DBO):
0
Subtask 3 (Violence):
0


<u><p>Interpretation</p></u>
No duplicate rows!

##### <b>Overlap between the subtask datasets</b>

In [18]:
datasets_to_compare = [
    ("Call2Action", df_call2action), 
    ("Violence", df_violence), 
    ("DBO", df_dbo)
]

# Calculates the raw overlaps
overlaps = data_utils.calculate_dataset_overlaps(
    datasets=datasets_to_compare,
    column="description",
    show_details=False
)

# Displays metrics
percentage_results = data_utils.display_dataset_overlap_percentages(
    overlaps=overlaps,
    datasets=datasets_to_compare,
    column="description",
    show_pairwise=True
)

Calculating overlaps across 3 datasets: Call2Action, Violence, DBO

Total unique descriptions appearing in at least two datasets: 264

───────────────────────────── Overlap Percentages for 'description' across 3 datasets ─────────────────────────────

Total overlap: 264/2868 unique descriptions appear in 2+ datasets (9.21%)

Pairwise percentages:

Call2Action and Violence share 100 descriptions:

- 100/1049 from Call2Action (9.53%)

- 100/1050 from Violence (9.52%)

- 100/1999 of all unique descriptions between them (5.00%)

Call2Action and DBO share 97 descriptions:

- 97/1049 from Call2Action (9.25%)

- 97/1049 from DBO (9.25%)

- 97/2001 of all unique descriptions between them (4.85%)

Violence and DBO share 99 descriptions:

- 99/1050 from Violence (9.43%)

- 99/1049 from DBO (9.44%)

- 99/2000 of all unique descriptions between them (4.95%)

<u><p>Interpretation:</p></u>

The analysis of overlapping tweet descriptions across the three subtask datasets (Call2Action, Violence, and DBO) reveals the following:

*   **Overall Corpus Overlap:**
    *   Out of a total of **2868 unique descriptions** across all three datasets combined, **264 descriptions (9.21%)** appear in *at least two* of these datasets.
    *   This indicates a moderate level of shared content when considering the entire corpus. A meaningful portion of the tweets is not unique to a single subtask's dataset.

*   **Pairwise Overlap Details:**

    *   **Call2Action vs. Violence:**
        *   Share **100** descriptions.
        *   These 100 descriptions represent **9.53%** of Call2Action's unique descriptions and **9.52%** of Violence's unique descriptions.
        *   Of the combined unique descriptions between *only* Call2Action and Violence (1999 descriptions), the 100 shared descriptions account for **5.00%**.
        *   -> A similar proportion of each dataset's content is shared with the other. The combined percentage suggests that while there's a direct overlap, the two datasets still largely contain distinct items from each other.

    *   **Call2Action vs. DBO:**
        *   Share **97** descriptions.
        *   These 97 descriptions represent **9.25%** of Call2Action's unique descriptions and **9.25%** of DBO's unique descriptions.
        *   Of the combined unique descriptions between *only* Call2Action and DBO (2001 descriptions), the 97 shared descriptions account for **4.85%**.
        *   -> Similar to the C2A-Violence pair, the overlap is proportionally balanced. The combined percentage again points to considerable distinct content between the two.

    *   **Violence vs. DBO:**
        *   Share **99** descriptions.
        *   These 99 descriptions represent **9.43%** of Violence's unique descriptions and **9.44%** of DBO's unique descriptions.
        *   Of the combined unique descriptions between *only* Violence and DBO (2000 descriptions), the 99 shared descriptions account for **4.95%**.
        *   -> Consistent pattern of balanced proportional overlap.

<u><p>Summary & Implications</p></u>

*   *Summary of Observations:*
    *   There's a consistent level of pairwise overlap (around 9-10% of each dataset involved in a pair is shared with the other).
    *   The "combined percentage" for each pair (around 5%) highlights that even when two datasets are considered together, the majority of their combined unique content is not shared.
    *   The overall 9.21% overlap across the entire corpus suggests that while subtasks might target different topics, some tweet instances are relevant to the annotation criteria of multiple subtasks.

*   *Implications*
    *   **Multi-Task Learning Potential:** The dataset overlap suggests that a **Multi-Task-Learning** ***(MTL)*** approach, where a single model predicts labels for multiple subtasks simultaneously, could be effective. This makes intuitive sense, as the overlap implies that the subtasks are not entirely distinct and likely share underlying characteristics. For instance, a tweet urging violent action to overthrow the German government could qualify as both a *'Call2Action'* and an *'Attack on DBO'*, indicating shared semantic features a single model could learn.
    *   **Data Leakage:** If datasets were to be merged into one dataset and then split into train/validation/test sets without considering these overlaps, identical tweets could contaminate the validation and test sets. This could lead to overly optimistic evaluation metrics for models that are effectively "seeing" test and validation data during training. In the case of merging the datasets, the dataset splits should be deduplicated to avoid the same text falling into more than one split (train, validation, or test).

##### <b>Analysing the overlapping tweets</b>

In [ ]:
df_overlapping_tweets = pd.DataFrame(list(overlaps["total"]), columns=["description"])

console.rule("[bold cyan]Overlapping Tweet Descriptions[/bold cyan]")
console.print(f"Total unique overlapping descriptions: {len(df_overlapping_tweets)}")

# Displays the first 3, last 3, and 3 random samples.
console.print("\n[bold]Head (first 5):[/bold]")
for tweet in df_overlapping_tweets["description"].head(5):
    console.print(f"- \"{tweet}\"")

console.print("\n[bold]Tail (last 5):[/bold]")
for tweet in df_overlapping_tweets["description"].tail(5):
    console.print(f"- \"{tweet}\"")

console.print("\n[bold]Random sample (5):[/bold]")
sample_size = min(5, len(df_overlapping_tweets))
for tweet in df_overlapping_tweets["description"].sample(sample_size, random_state=42):
    console.print(f"- \"{tweet}\"")

───────────────────────────────────────── Overlapping Tweet Descriptions ──────────────────────────────────────────

Total unique overlapping descriptions: 264

Head (first 5):

- "PEGIDA macht GLÜCKLICH !!!"

- "Frau Merkel wenn kennen Sie persönlich??   Einen Grapscher oder Halsabschneider??"

- "Sehr geehrter Ramon De la Perez, ich befürchte, dass Sie hier die Lage verkennen. Es handelt sich bei den so 
genannten Schutzsuchenden keineswegs um ein paar Jugendliche sondern um gewaltbereite junge Männer die sich 
scheinbar nicht scheuen sich das zu nehmen was sie wollen und Gesetze zu brechen. Anbei ein Link, der sicherlich 
für einen Träumer, wie sie es scheinbar sind, interessant ist: 
http://www.civitas-institut.de/index.php?option=com_content&view=article&id=2445%3Aerlebnisbericht-von-der-grenze-i
mmigranten-qhorden-von-wildenq&catid=1%3Aneuestes&Itemid=33
Míralo! Se comportaron como locos Y eso nos da miedo. No te lo crees que somos todos nazis. ¡Despierta te chico!!!"

- "Maxi guck deine Kinderserie??"

- "Wenn es schädlich ist, das eigene (deutsche Kräfte) wieder Arbeit erhalten und keine Ausländer, dann ist es doch
schön schädlich zu sein. Die Bundesanstalt für Arbeit ist wohl mit wenig bis kein Hirn gesegnet."

Tail (last 5):

- "Edward Jokop Sie haben vllt vergessen, dass Tillich schon seit 2009 im Amt ist - also wiedergewählt. Ihre 
Äußerungen sind dahingehend völlig verwirrend."

- "127 raus 3000 rein oh mein gott."

- "Wenn ein Vergewaltiger ohne Strafe davon kommt dann sieht aus und wahrscheinlich ER versteht es auch dass er das
RECHT hat. Vergewaltigung ist ein Verbrechen und soll streng bestraft werden."

- "Jetzt beleidigt und abhauen ... aber natürlich, wir sind unter deinem Niveau ;-) Schwach einfach, ganz schwach"

- "Johann Schaffer"

Random sample (5):

- "Es sieht wunderschön aus... Genau wie damals. Weiter so..."

- "Ich nenn die "arbeits"Agentur schädlich!!!!"

- "Für schlichte Charaktere gibt's halt die Simpsons..."

- "Wem soll man glauben???"

- "Schweine in jedes Heim !"

<u><p>Interpretation:</p></u>

We find, that the overlapping tweets consist of largely **substantive and meaningful tweets**, not merely collections of stopwords or noise.

*   **Observation:** The type of overlapping content ranges from short exclamations (e.g., `"Ihr seid soooooo böse"`, `"ja,und nun??"`) to longer statements (e.g., `"schau Dir die Kinder an. Die werden benutzt! Die sollte man rausholen, aber die Eltern dort lassen"`) and URLs.

*   **Key Implication:** ***Potential Data Leakage!*** Since these are actual, distinct tweet messages appearing in multiple datasets, their unmanaged presence across training, validation, and testing splits (if splits are done independently per subtask) could lead to models being evaluated on data they have effectively seen during training. This would compromise the integrity and meaningfulness of our performance metrics in the end.

*   **Conclusion:** Careful *filtering* and *data splitting* will likely be necessary to prevent our metrics from being clouded by tweets overlapping the datasets. (In an MTL setting)

##### <b>Basic Statistics: <i>Count, Unique, Top, Freq</i></b>

In [41]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action["description"].describe())
print(f"Subtask 2 (DBO):")
print(df_dbo["description"].describe())
print(f"Subtask 3 (Violence):")
print(df_violence["description"].describe())

Subtask 1 (Call2Action):
count     1051
unique    1049
top         ??
freq         2
Name: description, dtype: object
Subtask 2 (DBO):
count               1053
unique              1049
top       Danke Dresden!
freq                   2
Name: description, dtype: object
Subtask 3 (Violence):
count     1051
unique    1050
top         ??
freq         2
Name: description, dtype: object


<u><p>Interpretation</p></u>

*   **`count` (Number of Non-Missing Descriptions):**
    *   **C2A:** 1051 description field entries.
    *   **DBO:** 1053 description field entries.
    *   **VIO:** 1051 description field entries.
    *   *Observation:* Matches the total row counts for each subtask. No missing tweet descriptions.

*   **`unique` (Number of Distinct Descriptions):**
    *   **C2A:** 1049 unique tweet texts.
        *   *Observation:* Two duplicate tweets.
    *   **DBO:** 1049 unique tweet texts.
        *   *Observation:* Four duplicate tweets.
    *   **VIO:** 1050 unique tweet texts.
        *   *Observation:* 1 duplicate tweet.

*   **`top` (Most Frequent Description):**
    *   *Info:* Doesn't hold significant added information, as the amount of most frequent descriptions is small.

*   **`freq` (Frequency of the Top Description):**
    *   *Info:* Shows that the `top` tweet in each subtask appears 2 times, aligning with the small number of duplicates identified.

<u><p>Summary</p></u>

*   *Observation:* The trial datasets for each subtask contain a very small number of duplicate tweet texts (1 to 4 instances per ~1050 tweets). The vast majority of descriptions are unique.